### PR TITLE
url based acrs and unescaped whitespace in URI

### LINKED ISSUE #345: URL based Stepped up authentication in OIDC plugin
Add feature in the OIDC plugin that enables stepped up authentication based on URL (resource), e.g.: 

1. `site.com/account_info` –> prompt=login acr=basic
2. `site.com/wire_transfer` –> prompt=login acr=super_gluu

### LINKED ISSUE #347: Unescaped space character in a URI - possible security vulnerabilities
Most of the web servers would ignore the URI suffix after unescaped whitespace character.

We use `ngx.var.uri` to obtain the normalized path.
Normalized means that any relative path are resolved and query parameters removed.

But `ngx.var.uri` would return all chars after unescaped space also and our matching code may fail.
In the worst key, we may wrongly provide access to a protected resource.

IMO we need to strip all chars after first unescaped whitespace before URI matching.
It is a quick fix.

Plugins:
- gluu-oauth-pep
- gluu-oauth-uma
- gluu-openid-connect

### PR DESCRIPTION
**Pull request by altexy**
Close #347 
Close  #345

### CODE REVIEW

**Comment by ldeveloperl1985** [comment]
Hi @altexy ,

Tested with real Gluu and OXD, getting a problem `{"message":"Authentication Context Class is not enough"}`. I am trying to stepped up auth for `/users/??` endpoint and for other basic auth.

I've registered the below resources
```
"required_acrs_expression": [
      {
        "path": "/users/??",
        "conditions": [
          {
            "required_acrs": [
              "otp"
            ],
            "httpMethods": [
              "?"
            ]
          }
        ]
      },
      {
        "path": "/??",
        "conditions": [
          {
            "required_acrs": [
              "auth_ldap_server"
            ],
            "httpMethods": [
              "?"
            ]
          }
        ]
      }
    ],
```
```
    "required_acrs": [
      "auth_ldap_server",
      "otp"
    ]
```
```
"method_path_tree": {
      "?": {
        "users": {
          "??": {
            "#": {
              "path": "/users/??",
              "required_acrs": [
                "otp"
              ]
            }
          }
        },
        "??": {
          "#": {
            "path": "/??",
            "required_acrs": [
              "auth_ldap_server"
            ]
          }
        }
      }
    },
```

Reproduce: 
1. First request, URL `https://kong.proxy` --> here it only basic auth happened and able to access resources. here, i've checked the id token acr values are `  "acr": "auth_ldap_server",`
2.  Second request, URL `https://kong.proxy` --> getting `{"message":"Authentication Context Class is not enough"}` problem

Please check kong logs, from DateTime `2019/08/06 17:05:48`
[error.txt](https://github.com/GluuFederation/gluu-gateway/files/3472015/error.txt)

I think the problem is here, not sure please check. https://github.com/GluuFederation/gluu-gateway/blob/09ac1968f0bf5b13d51b71a1e52908daf169a06e/kong/plugins/gluu-openid-connect/access.lua#L211.

**Comment by willow9886** [comment]
> Authentication Context Class is not enough 

This error message is a bit cryptic as well.. what does that mean?

**Comment by altexy** [comment]
@ldeveloperl1985 
> I think the problem is here, not sure please check.

yes, you are right, thanks, fixed

**Comment by altexy** [comment]
@willow9886 
>> Authentication Context Class is not enough
> This error message is a bit cryptic as well.. what does that mean?

`Authentication Context Class` is a term of [OpenID Connect core spec](https://openid.net/specs/openid-connect-core-1_0-final.html#Terminology)
It is authentication methods, for example `opt`, `super_gluu`, `u2f`

We may be more detailed: specify acrs the user has and the acrs required for the resource.

**Comment by ldeveloperl1985** [comment]
@altexy I just tested the flow but after 2FA auth when I click on other paths, it again goes for basic auth.

Reproduce:
1. request `https://kong-proxy.com/home` --> asking for basic auth and allow
1. request `https://kong-proxy.com/users/` --> asking for 2FA auth and allow which is ok
1. but when I request again to `https://kong-proxy.com/home`, it again asking for basic auth.

**Comment by altexy** [comment]
> @altexy I just tested the flow but after 2FA auth when I click on other paths, it again goes for basic auth.

It looks like GS doesn't support multiple acrs for a user.
We may maintain multiple id_tokens in user session, not sure would these tokens all valid from GS point of view.

I will implement support of multiple id_tokens in user session asap, will see.

**Comment by willow9886** [comment]
> @willow9886
> 
> > > Authentication Context Class is not enough
> > > This error message is a bit cryptic as well.. what does that mean?
> 
> `Authentication Context Class` is a term of [OpenID Connect core spec](https://openid.net/specs/openid-connect-core-1_0-final.html#Terminology)
> It is authentication methods, for example `opt`, `super_gluu`, `u2f`
> 
> We may be more detailed: specify acrs the user has and the acrs required for the resource.

Yes, I'm familiar with OpenID Connect ACR values. The error message is not descriptive enough to explain the issue. `...is not enough`.. what?

**Comment by yuriyz** [comment]
Two notes:

- Shouldn't GG rely on acr level ? If we have lets say `fido2` (70) then we don't need `basic` (20). It should be accepted automatically. I guess that's was the original idea when levels of acr were introduced. What is the point of weaker authentication?
```
"auth_level_mapping" : {
    "-1" : [ "auth_ldap_server" ],
    "70" : [ "fido2" ],
    "20" : [ "basic_lock" ],
    "10" : [ "basic" ]
  },
```

2. Re: `Authentication Context Class is not enough.` it probably should sound like `Required ACR value is not found. Stepped up authentication is needed. Requires acr: ["fido2"]`. In this case it gives hint how to resolve the problem.

File: kong/plugins/gluu-openid-connect/access.lua [hunk 1]
```diff
@@ -198,6 +226,7 @@ local function authorize(conf, session, prompt, required_acrs)
         return unexpected_error()
     end
 
+    -- TODO does oxd support this now?
```
Comment (ldeveloperl1985) [inline kong/plugins/gluu-openid-connect/access.lua]: Yes, It is available now in OXD https://gluu.org/docs/oxd/4.0/api/#get-authorization-url

File: kong/plugins/gluu-openid-connect/access.lua [hunk 2]
```diff
@@ -252,45 +301,90 @@ return function(self, conf)
...
         local rule = path_wildcard_tree.matchPath(method_path_tree, ngx.req.get_method(), path)
         required_acrs = rule and rule.required_acrs
     end
-    required_acrs =  required_acrs or conf.required_acrs
+    required_acrs =  required_acrs or conf.required_acrs -- TODO remove conf.required_acrs here should be default exp
+
+    if required_acrs and #required_acrs == 0  then
+        -- treat empty array as no auth required
+        -- missed required_acrs treated as any acr as before
+        return
```
Comment (ldeveloperl1985) [inline kong/plugins/gluu-openid-connect/access.lua]: As we discussed, I agree with `missed required_acrs --->  no-auth required for a 
 resource`
About the `Empty ACR case`,  IMO, the user must have to add at least one ACR for the protected path, so no more empty ACR case.

Comment (ldeveloperl1985) [reply kong/plugins/gluu-openid-connect/access.lua]: Also, we removed `conf.required_acrs` from schema.lua so at least need one acr for the protected path.

**Pull request merged**